subsys/bluetooth/mesh/adv_ext.c

```diff
@@ -314,47 +341,37 @@ static void send_pending_adv(struct k_work *work)
 		}
 	}
 
-	while ((adv = bt_mesh_adv_get_by_tag(ext_adv->tags, K_NO_WAIT))) {
-		/* busy == 0 means this was canceled */
-		if (!adv->ctx.busy) {
-			bt_mesh_adv_unref(adv);
-			continue;
+	do {
+		sent = adv_send_process(adv, ext_adv);
+		if (sent) {
+			return;
 		}
 
-		adv->ctx.busy = 0U;
-		err = adv_send(ext_adv, adv);
-
-		bt_mesh_adv_unref(adv);
-
-		if (!err) {
-			return; /* Wait for advertising to finish */
+		if (ext_adv->instance == NULL) {
+			LOG_DBG("Advertiser is suspended or deleted");
+			return;
 		}
-	}
-
-	if (ext_adv->instance == NULL) {
-		LOG_DBG("Advertiser is suspended or deleted");
-		return;
-	}
-
-	if (IS_ENABLED(CONFIG_BT_MESH_PROXY_SOLICITATION) &&
-	    !bt_mesh_sol_send()) {
-		return;
-	}
 
-	if (!IS_ENABLED(CONFIG_BT_MESH_GATT_SERVER) ||
-	    !(ext_adv->tags & BT_MESH_ADV_TAG_BIT_PROXY)) {
-		return;
-	}
+		if (IS_ENABLED(CONFIG_BT_MESH_PROXY_SOLICITATION) &&
+		    !bt_mesh_sol_send()) {
+			return;
+		}
 
-	atomic_set_bit(ext_adv->flags, ADV_FLAG_PROXY_START);
+		if (!IS_ENABLED(CONFIG_BT_MESH_GATT_SERVER) ||
+		    !(ext_adv->tags & BT_MESH_ADV_TAG_BIT_PROXY)) {
+			return;
+		}
 
-	if (!bt_mesh_adv_gatt_send()) {
-		atomic_set_bit(ext_adv->flags, ADV_FLAG_PROXY);
-	}
+		if (!bt_mesh_adv_gatt_send()) {
+			atomic_set_bit(ext_adv->flags, ADV_FLAG_PROXY);
+		}
 
-	if (atomic_test_and_clear_bit(ext_adv->flags, ADV_FLAG_SCHEDULE_PENDING)) {
-		schedule_send_with_mask(ext_adv, K_WORK_RUNNING);
-	}
+		/* If there is a pending adv during the process of starting the
+		 * proxy advertising, the proxy advertising needs to be stop and
+		 * sent pending advertising immediately.
+		 */
+	} while ((adv = bt_mesh_adv_get_by_tag(ext_adv->tags, K_NO_WAIT)) &&
```

Comment:
I don't want to use rescheduling anymore because it will inevitably add extra flags, which is clearly contrary to the original intention of my PR

Additionally, this example is too extreme and I believe it lacks practical significance.

Under normal circumstances, adv_dend_process should be sent successfully. Even if it fails to send, it should not continue to fail. For continuous sending failures, this is already a problem. It is meaningless to discuss how to avoid endless loops in this situation.

Reply:
I can't agree that this is extreme. Now, we the changed host where it can return ENOBUF, this becomes an issue when running the mesh ext advertiser on sysworkq. If there is a thread competing with mesh for the host's buffers, the scenario I explained above will happen.

Reply:
Well, compromise for this reason, add `k_yield`.